Realm/RLMSyncSubscription.h

```diff
@@ -0,0 +1,133 @@
+// Copyright 2018 Realm Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+
+#import <Realm/RLMResults.h>
+
+NS_ASSUME_NONNULL_BEGIN
+
+/**
+ `RLMSyncSubscriptionState` is an enumeration representing the possible state of a sync subscription.
+ */
+typedef NS_ENUM(NSInteger, RLMSyncSubscriptionState) {
+    /**
+     An error occurred while creating the subscription or while the server was processing it.
+     */
+    RLMSyncSubscriptionStateError = -1,
+
+    /**
+     The subscription is being created, but has not yet been written to the synced Realm.
+     */
+    RLMSyncSubscriptionStateCreating = 2,
+
+    /**
+     The subscription has been created, and is waiting to be processed by the server.
+     */
+    RLMSyncSubscriptionStatePending = 0,
+
+    /**
+     The subscription has been processed by the server, and objects matching the subscription
+     are now being synchronized to this client.
+     */
+    RLMSyncSubscriptionStateComplete = 1,
+
+    /**
+     This subscription has been removed.
+     */
+    RLMSyncSubscriptionStateInvalidated = 3,
+};
+
+/**
+ `RLMSyncSubscription` represents a subscription to a set of objects in a synced Realm.
+
+ When partial sync is enabled for a synced Realm, the only objects that the server synchronizes to the
+ client are those that match a sync subscription registered by that client. A subscription consists of
+ of a query (represented by an `RLMResults`) and an optional name.
+
+ The state of the subscription can be observed using [Key-Value Observing](https://developer.apple.com/library/content/documentation/Cocoa/Conceptual/KeyValueObserving/KeyValueObserving.html) on the `state` property.
+
+ Subscriptions are created using `-[RLMResults subscribe]` or `-[RLMResults subscribeWithName:]`.
+ */
+@interface RLMSyncSubscription : NSObject
+
+/**
+ The unique name for this subscription.
+
+ This will be `nil` if a name was not provided when the subscription was created.
+ */
+@property (nonatomic, readonly, nullable) NSString *name;
+
+/**
+ The state of the subscription. See `RLMSyncSubscriptionState`.
+ */
+@property (nonatomic, readonly) RLMSyncSubscriptionState state;
+
+/**
+ The error associated with this subscription, if any.
+
+ Will be non-nil only when `state` is `RLMSyncSubscriptionStateError`.
+ */
+@property (nonatomic, readonly, nullable) NSError *error;
+
+/**
+ Remove this subscription.
```

Comment:
This could use a description of what happens when a subscription is removed. A rough take:

Removing a subscription will delete all objects from the local Realm which were matched only by that subscription and not any remaining subscriptions. This happens asynchronously following contact with the server, and has no immediate impact on the contents of the local Realm.